OpenTrons/opentrons-api
# app(feat): launch drop tip wizard logic

Closes RAUT-795



# Overview

This PR adds logic for wiring up the drop tip modal in additional locations. It also refactors the nice-modal-react context providers to the appropriate app/ODD locations to take advantage of top-level context. It includes some drive by refactoring as well - primarily adding missing tests.

A few notes of interest/controversy:
1) Because the OT-2 provides no way to validate that a tip has been removed, and we probably shouldn't always assume the Flex sensor is accurate for determining whether a tip is attached, the banner/ODD modal logic assumes that once a user begins the drop tip flow that the tip will be removed or the user is no longer interested with tip removal (and therefore the alert should be removed). 
2) If a user has tips on both the left and right pipettes and chooses not to skip the drop tip wizard (clicks dismiss on the dekstop banner or presses skip on the ODD), BUT the user exists a drop tip wizard flow prematurely, the banner/modal popup will still exist for the second pipette. 
3) The one optimization for getPipettesWithTipAttached is if both pipettes are used in a run, check for tip status for both pipettes on the same command loop iteration. 

### Protocol Header Entrypoint 
<img width="922" alt="Screenshot 2023-10-20 at 11 18 29 AM" src="https://github.com/Opentrons/opentrons/assets/64858653/8a532328-535a-4d69-9d8c-1e53092317b2">

### ODD InstrumentDetail Entrypoint
<img width="728" alt="Screenshot 2023-10-20 at 11 18 09 AM" src="https://github.com/Opentrons/opentrons/assets/64858653/b69e6d57-63c7-4080-9da6-2f9691648baf">

### ODD RunSummary Entrypoint

<img width="509" alt="Screenshot 2023-10-20 at 11 17 48 AM" src="https://github.com/Opentrons/opentrons/assets/64858653/fc92d8f4-7705-46a0-b45a-10d3257a5212">



# Test Plan

Protocols That May Be Useful:
[pick_up_tip.py.zip](https://github.com/Opentrons/opentrons/files/13055798/pick_up_tip.py.zip)
[Just Pipettes.py.zip](https://github.com/Opentrons/opentrons/files/13055797/Just.Pipettes.py.zip)

- Place tips on one/both of the pipettes and run the just pipettes protocol. After the run the banner should appear (Desktop) or clicking return to dashboard/run again will open the modal (ODD).
- Dismissing the banner will cause the drop tip modal not to render if it is not clicked in a few seconds. Clicking "skip" on the ODD will cause the drop tip banner on the desktop app to disappear. 
- Run the pick up tip protocol and press the estop some time AFTER the tip has attached. Verify the banner/ODD modal appears. 
- Run the pick up tip protocol and press the estop BEFORE the tip has attached (but before the pickUpTip command has been performed). Verify there is no banner/ODD modal.


# Changelog
- Added ODD overflow menu on the InstrumentDetail page.
- Wired up the drop tip wizard on the Desktop & ODD apps.


# Review requests
Test on a physical robot, please! AppUI is configured for these protocols.


# Risk assessment
low
